libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,717 @@
+        response_payload->burst_complete = false;
+        errorCode = _ftp_Burst(payload, target_system_id);
+        break;
+
+    case kCmdWriteFile:
+        errorCode = _ftp_Write(payload, response_payload);
+        break;
+
+    case kCmdRemoveFile:
+        errorCode = _ftp_RemoveFile(payload, response_payload);
+        break;
+
+    case kCmdRename:
+        errorCode = _ftp_Rename(payload, response_payload);
+        break;
+
+    case kCmdTruncateFile:
+        errorCode = _ftp_TruncateFile(payload, response_payload);
+        break;
+
+    case kCmdCreateDirectory:
+        errorCode = _ftp_CreateDirectory(payload, response_payload);
+        break;
+
+    case kCmdRemoveDirectory:
+        errorCode = _ftp_RemoveDirectory(payload, response_payload);
+        break;
+
+    case kCmdCalcFileCRC32:
+        errorCode = _ftp_CalcFileCRC32(payload, response_payload);
+        break;
+
+    default:
+        errorCode = kErrUnknownCommand;
+        break;
+    }
+
+out:
+    response_payload->seq_number = payload->seq_number + 1;
+
+    // handle success vs. error
+    if (errorCode == kErrNone) {
+        response_payload->req_opcode = payload->opcode;
+        response_payload->opcode = kRspAck;
+    }
+    else {
+        response_payload->req_opcode = payload->opcode;
+        response_payload->opcode = kRspNak;
+        response_payload->size = 1;
+        response_payload->data[0] = errorCode;
+        if (errorCode == kErrFailErrno) {
+            response_payload->size = 2;
+            response_payload->data[1] = _ftp_errno;
+        }
+    }
+
+    //Send Response
+    _ftp_response.target_system = target_system_id;
+    mavlink_msg_file_transfer_protocol_send_struct(chan, &_ftp_response);
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_List(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    FTPErrorCode errorCode = kErrNone;
+    unsigned offset = 0;
+    char* dir_name;
+    _set_current_abs_path(_data_as_cstring(request_payload));
+    dir_name = _ftp_current_abs_path;
+    DIR *dp = opendir(dir_name);
+
+    if (dp == nullptr) {
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "FTP: can't open path (file system corrupted?)");
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), dir_name);
+        _ftp_errno = errno;
+        // this is not an FTP error, abort directory by simulating eof
+        return kErrEOF;
+    }
+
+    struct dirent *result = nullptr;
+
+    // move to the requested offset
+    int requested_offset = request_payload->offset;
+    response_payload->offset = request_payload->offset;
+    while (requested_offset-- > 0 && readdir(dp));
+
+    while ((offset + 2) < MAX_FTP_DATA_LENGTH) {
+        errno = 0;
+        result = readdir(dp);
+        // read the directory entry
+        if (result == nullptr) {
+            if (errno) {
+                gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "FTP: list readdir failure");
```

Comment:
this looks like developer debugging? Maybe add a debug() macro and put a way to enable debugging as an ifdef at top of file?